design/14_tinkerbell_crd_refactor/proposal.md

```diff
@@ -0,0 +1,462 @@
+	Args []string
+
+	// Env defines environment variables used when launching the container.
+	// Optional.
+	Env map[string]string
+
+		// Volumes defines the volumes to mount into the container.
+	// Optional.
+	Volumes []Volume
+
+	// NetworkNamespace defines the network namespace to run the container in. This enables access 
+	// to the host network namespace.
+	// See https://man7.org/linux/man-pages/man7/namespaces.7.html.
+	NetworkNamespace string
+}
+
+// Volume is a specification for mounting a volume in an action. Volumes take the form
+// {SRC-VOLUME-NAME | SRC-HOST-DIR}:TGT-CONTAINER-DIR:OPTIONS. When specifying a VOLUME-NAME that 
+// does not exist it will be created for you.
+//
+// Examples
+//
+// Read-only bind mount bound to /data
+//   /etc/data:/data:ro
+//
+// Writable volume name bound to /data
+//   shared_volume:/data
+//
+// See https://docs.docker.com/storage/volumes/ for additional details
+type Volume string
+
+// ActionName is unique name within the context of a workflow.
+type ActionName string
+```
+
+#### `Workflow`
+
+```go
+// Workflow describes a set of actions to be run on a specific Hardware. Workflows execute
+// once and should be considered ephemeral.
+type Workflow struct {
+	Spec   WorkflowSpec
+	Status WorkflowStatus
+}
+
+type WorkflowSpec struct {
+	// HardwareRef is a reference to a Hardware resource this workflow will execute on.
+	// If no namespace is specified the Workflow's namespace is assumed.
+	HardwareRef LocalObjectReference
+
+	// TemplateRef is a reference to a Template resource used to render workflow actions.
+	// If no namespace is specified the Workflow's namespace is assumed.
+	TemplateRef LocalObjectReference
+
+	// TemplateData is user defined data that is injected during template rendering. The data
+	// structure should be marshalable.
+	// Optional.
+	TemplateData map[string]any
+
+	// Timeout defines the time the workflow has to complete. The timer begins when the first action
+	// is requested. When set to 0, no timeout is applied.
+	// Default 0.
+	Timeout int32
+}
+
+type WorkflowStatus struct {
+	// Actions is the list of rendered actions and their status.
+	Actions RenderedActions
+
+	// StartedAt is the time the first action was requested. Nil indicates the workflow has not
+	// started.
+	StartedAt *metav1.Time
+
+	// State describes the current state of the workflow. This fields represents a summation of
+	// action states. Specifically, if all actions succeeded, the workflow will succeed. If one
+	// action fails, the workflow fails irrespective of previous action status'.
+	State State
+
+	// Reason describes the reason for failure. It is only relevant when Result is ResultFailed.
+	// It is propogated from the failed action.
+	Reason Reason
+}
+
+// RenderedActions is a map of action name to RenderedAction.
+type RenderedActions map[ActionName]ActionStatus
+
+// ActionStatus describes status information about an action.
+type ActionStatus struct {
```

Comment:
I would strongly recommend of add the conditions framework here instead of using explicit fields like State. Something similar to this https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/condition_types.go

Reply:
I flip-flopped on whether to use conditions. Here's my train of thought:

1. Conditions are meant to be observations, not state machines in and of themselves. I think this is a common misinterpretation and misuse. Cluster API has a good practical example with its `ControlPlaneReady` condition. `ControlPlaneReady` transition to `True` when the first control plane node is up, but it can also transition back to `False` (at least that's my understanding). Conditions can be monotonic but, again, that doesn't mean they should represent state machines. [The guidance](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) is that they shouldn't represent state machines.

2. Adding conditions to actions creates a set of nested conditions because the `Workflow` would have conditions which are a summarization of `ActionStatus` conditions. I don't think I'm totally opposed to this approach, but I fear its confusing given how established conditions are in Kubernetes and the fact they generally refer to the Kind, not nested data structures of a Kind. We could add conditions to the `WorkflowStatus` and derive them from `ActionStatus` information; that feels more appropriate given all other guidance but they would be complimentary to state machine fields as opposed to the actual representation of the state machine.

3. Conditions _should_ generally be derivable from information already available on the resource. That is by no means a hard and fast rule but encoding state into conditions that need to be populated at reconcile time means state expectations are hard to rationalize when looking at the API only (documentation may clear this up with ease, I haven't given it much thought).

4. Conditions have a [predefined recommended format](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) which doesn't necessarily compliment what we're doing. We could break from this in the same way Cluster API has (they dropped the `ObservedGeneration` field) and that's probably fine but it contributes to the [turmoil](https://maelvls.dev/kubernetes-conditions/) already shrouding conditions. We would need to very clearly, [as Cluster API has](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200506-conditions.md), define our condition expectations.

---

What does it look like if we used conditions? I would probably forgo most of the 'states' we have in favor of the following condition types.

1. `Started`: A monotonic condition indicating the workflow has begun.
2. `Succeeded`: A monotonic condition indicating the workflow succeeded. `Unknown` indicates it hasn't finished and `False` indicates it failed. On `False` I would expect a `Reason` and `Message` elaborating on why it didn't succeed.

Reply:
I think this might be useful to document in the proposal's `Rationale` section, whatever we decide.

Reply:
These are quite interesting thoughts I would say. Ok maybe we can move the usage of conditions in another PR/Proposal? One thing to keep in mind though is, we should distinguish what things are relevant to the CR itself and it is being reconciled and how the underlying resource(machine) has eventually turned out to be. 

I have to say though that, I have seen multiple cases, where a status of an object is being derived from another status assigned of another object. I mean for now I guess we can roll without it, and then we can iterate in the future about what's needs to be done here :)

Reply:
I've added the detail to the proposals rationale section and said we will tackle conditions as a separate proposal.

Reply:
@moadqassem I thought some more on conditions, here's what I've come up with.

1. We'll have the `Reason` and `Message` fields on `ActionStatus` as detailed in the proposal already.
2. As part of this proposal, we'll define how action developers can propagate a `Reason` and `Message` from within actions such that it populates the `ActionStatus` fields.
3. We'll remove `Reason` and `Message` from `WorkflowStatus` and replace it with a conditions data structure (the conditions will include a `Reason` and `Message`).
4. We'll define a `Successful` condition. The values of `Reason` and `Message` will be propagated from the failed `ActonStatus` providing the summarized observation of the workflow.

This does mean we'll need to define conditions as part of this proposal. Given we know we want to provide ways for users to propagate information from within actions on errors, and that we want to trend toward conditions anyway, it feels appropriate to just go ahead and define our conditions setup as part of this proposal. 

Using conditions on the `WorkflowStatus` to summarize action/workflow state feels more extensible and avoids duplicating the information field for field  between `ActionStatus` and `WorkflowStatus`.

Conditions will likely take the [same form as Cluster APIs](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200506-conditions.md#data-model-changes).

The TL;DR on `Reason` and `Message` propagation from within actions is to specify paths such as `/tinkernell/reason` and `/tinkerbell/message` that action developers can write to when they fail. We'll read these files and propagate the strings from client to server which will update the custom resource.

We could optionally define a `Started` condition which is another easily obtained observation that could be helpful for snapshotting resource state.

Reply:
After discussing with Jacob, I decided to shoot for this ahead of tomorrow's community call. The proposal has been updated.